EIPS/eip-4361.md

```diff
@@ -0,0 +1,295 @@
+---
+eip: 4361
+title: Sign-In with Ethereum. 
+description: Standardizes off-chain authentication for Ethereum accounts to establish sessions.
+author: Wayne Chang (@wyc), Brantly Millegan (@brantlymillegan), Nick Johnson (@Arachnid)
+resolution: <TBD>
+discussions-to: https://ethereum-magicians.org/t/eip-4361-sign-in-with-ethereum/7263
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-10-11
+requires: 55, 191, 137, 155, 1271, 1328
+---
+
+## Abstract
+Sign-In with Ethereum describes how Ethereum accounts authenticate with off-chain services by signing a standard message format parameterized by scope, session details, and security mechanisms (e.g., a nonce). The goals of this specification are to provide a self-custodied alternative to centralized identity providers, improve interoperability across off-chain services for Ethereum-based authentication, and provide wallet vendors a consistent machine-readable message format to achieve improved user experiences and consent management.
+
+## Motivation
+When signing in to popular non-blockchain services today, users will typically use identity providers (IdPs) that are centralized entities with ultimate control over users' identifiers, for example, large internet companies and email providers. Incentives are often misaligned between these parties. Sign-In with Ethereum offers a new self-custodial option for users who wish to assume more control and responsibility over their own digital identity.
+
+Already, many services support workflows to authenticate Ethereum accounts using message signing, such as to establish a cookie-based web session which can manage privileged metadata about the authenticating address. This is an opportunity to standardize the sign-in workflow and improve interoperability across existing services, while also providing wallet vendors a reliable method to identify signing requests as Sign-In with Ethereum requests for improved UX.
+
+This work is sponsored by the Ethereum Foundation and Ethereum Name Service (ENS). It is being developed in the open through a series of recorded community calls and public repositories, and its development is informed by over twenty user interviews with a focus on currently-in-production uses, related prior EIPs, and fits within product roadmaps.
+
+## Specification
+Sign-In with Ethereum works as follows:
+
+1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the EIP-191 signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.
```

Comment:
I tend to agree with @PhABC here. It's not completely obvious the requirement for a `nonce` and what benefit nonce actually providers for "the most developers" category.

Sure, for med/high security situations, it's important to re-authenticate a user before allowing roles based access control to resources (compute or otherwise), but I would argue "generic" authentication uses a timestamps for session management and if a re-authentication is required you invalidate the active session signature/JWT

Could easily see most services constantly using the `0` or `1` nonce, because they have too, but don't have a real need for it.

If a server is going to be "stateful it would be good to research what's the most common techniques for invalidation of active sessions. To my knowledge a lot services will just invalidate a previous session cookie and require a user re-authenticate, which nullifies the requirement for a nonce anyways, because they can be handled via timestamp parameters.

Reply:
Hmm maybe, I'm missing something here but can someone explain why login is necessary in a stateless app interaction? I want to make sure I understand the architecture of how this is being used since I think I may be missing the point in why this would be useful.

Reply:
@kdenhartog I think by stateless what's being expressed is no database in the authentication layer. Traditionally with a signed JWT you can get away with authenticating signatures without pinging a database to validate a session. If the signature is valid and the timestamp hasn't expired it's good to go.

Reply:
>Traditionally with a signed JWT you can get away with authenticating signatures without pinging a database to validate a session. If the signature is valid and the timestamp hasn't expired it's good to go.

Yes, but that's because the resource server has an inherent binding to the authorization server because it's directly retrieving the token from the OP in the authorization code flow. So in other words the application itself can be managed in a stateless manner because the OP can be trusted to not maliciously replay the token due to the trust between the two servers in that architecture. This means the nonce is unnecessary to be used for replay protection and it's only being used for the second purpose which is to associate a Client session with an ID Token. However, because the authorization code is also being used to associate the client session with the ID Token, the nonce no longer is necessary in this flow.

The issue here is that the resource server can't trust the authorization server in this case and so the architecture of SIWE is more in line with the implicit code flow which does require the usage of a nonce to be sent in the request. This is in section [3.2.2.1 of the OIDC spec](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitAuthRequest).

Building on this the spec says in section [3.2.2.11 of the OIDC Core spec](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitIDTValidation) it states:

> if a nonce value was sent in the Authentication Request, a nonce Claim MUST be present and its value checked to verify that it is the same value as the one that was sent in the Authentication Request. The Client SHOULD check the nonce value for replay attacks. The precise method for detecting replay attacks is Client specific. 

So in other words, because there's not an inherent trust between the RP and the wallet a client MUST be used. If an app wanted to go stateless with the current design they would most likely need to utilize the OAuth server that Spruce also has committed to so that the RP could establish a trust and defer authn and authz events to the OP.

Reply:
Sorry... I'm a little confused. What resource server? At the end of the day of this is minimally self-signed JWT-like token that's rooted to external sources via a verifiable address.

What are we getting from adopting that OIDC spec when we're removing the IDP and RP from the equation?

Reply:
Hmmm... yeah I just don't see why that matters with the self-signing of the authentication token. The trust is from signature verification and `issuedAt` key, which essentially acts as `nonce` so why add a field to semi-adopt the ODIC spec?

Reply:
It's not about trying to adopt the OIDC spec. I only used that as an example in order to showcase this is standard for protocol designs like this and to highlight that trust boundaries between the different software systems matter here. In this particular case due to the nature in which the trust works, the appropriate method to be used is to require the nonce in order to allow the protocol to be used in a broader set of use cases without requiring dApp developers to have to assess their threat model each time and decide if it's necessary in order to support 1 architecture which I'm still not convinced makes legitimate sense.

Reply:
Thanks for your comments, if you continue to feel strongly about this please continue the discussion in ETH Magicians.